File: sdk/packages/python/iii/src/iii/iii_constants.py
"""Constants and configuration types for the III SDK (mirrors iii-constants.ts)."""

from dataclasses import dataclass
from typing import Any, Callable, Literal

from .telemetry_types import OtelConfig

IIIConnectionState = Literal["disconnected", "connecting", "connected", "reconnecting", "failed"]

ConnectionStateCallback = Callable[["IIIConnectionState"], None]

DEFAULT_INVOCATION_TIMEOUT_MS = 30000
MAX_QUEUE_SIZE = 1000
DEFAULT_MAX_MESSAGE_SIZE = 16 * 1024 * 1024


@dataclass
class ReconnectionConfig:
    """Configuration for WebSocket reconnection behavior.

    Attributes:
        initial_delay_ms: Starting delay in milliseconds. Default ``1000``.
        max_delay_ms: Maximum delay cap in milliseconds. Default ``30000``.
        backoff_multiplier: Exponential backoff multiplier. Default ``2.0``.
        jitter_factor: Random jitter factor (0--1). Default ``0.3``.
        max_retries: Maximum retry attempts. ``-1`` for infinite. Default ``-1``.
    """

    initial_delay_ms: int = 1000
    max_delay_ms: int = 30000
    backoff_multiplier: float = 2.0
    jitter_factor: float = 0.3
    max_retries: int = -1


DEFAULT_RECONNECTION_CONFIG = ReconnectionConfig()


@dataclass
class FunctionRef:
    """Reference to a registered function, allowing programmatic unregistration."""

    id: str
    unregister: Callable[[], None]


@dataclass
class TelemetryOptions:
    """Telemetry metadata to be reported to the engine.

    Attributes:
        language: Programming language of the worker (e.g. ``python``).
        project_name: Name of the project this worker belongs to.
        framework: Framework name (e.g. ``motia``) if applicable.
        amplitude_api_key: Amplitude API key for product analytics.
    """

    language: str | None = None
    project_name: str | None = None
    framework: str | None = None
    amplitude_api_key: str | None = None


@dataclass
class InitOptions:
    """Options for configuring the III SDK.

    Attributes:
        worker_name: Display name for this worker. Defaults to ``hostname:pid``.
        enable_metrics_reporting: Enable worker metrics via OpenTelemetry. Default ``True``.
        invocation_timeout_ms: Default timeout for ``trigger()`` in milliseconds. Default ``30000``.
        reconnection_config: WebSocket reconnection behavior.
        otel: OpenTelemetry configuration. Enabled by default.
            Set ``{'enabled': False}`` or env ``OTEL_ENABLED=false`` to disable.
        telemetry: Internal telemetry metadata.
        max_message_size: Maximum size in bytes of a single WebSocket message
            sent to or received from the engine. Default ``16 MiB``. Direct
            ``trigger()`` payloads ride a single message; for larger or
            streamable payloads use channels. The engine enforces its own
            ceiling — keep this in sync with the engine config.
    """

    worker_name: str | None = None
    enable_metrics_reporting: bool = True
    invocation_timeout_ms: int = DEFAULT_INVOCATION_TIMEOUT_MS
    reconnection_config: ReconnectionConfig | None = None
    otel: OtelConfig | dict[str, Any] | None = None
    headers: dict[str, str] | None = None
    telemetry: TelemetryOptions | None = None
    max_message_size: int = DEFAULT_MAX_MESSAGE_SIZE

File: sdk/packages/node/iii-browser/src/iii-constants.ts
/**
 * Constants for the III module.
 */

/**
 * Engine function paths for internal operations.
 *
 * Naming note: `LIST_TRIGGERS` / `INFO_TRIGGERS` cover trigger TYPES
 * (templates). `LIST_REGISTERED_TRIGGERS` / `INFO_REGISTERED_TRIGGERS`
 * cover trigger INSTANCES (subscriber rows). The old
 * `engine::trigger-types::list` builtin has been removed and is now
 * served by `engine::triggers::list`.
 */
export const EngineFunctions = {
  LIST_FUNCTIONS: 'engine::functions::list',
  INFO_FUNCTIONS: 'engine::functions::info',
  LIST_WORKERS: 'engine::workers::list',
  INFO_WORKERS: 'engine::workers::info',
  LIST_TRIGGERS: 'engine::triggers::list',
  INFO_TRIGGERS: 'engine::triggers::info',
  LIST_REGISTERED_TRIGGERS: 'engine::registered-triggers::list',
  INFO_REGISTERED_TRIGGERS: 'engine::registered-triggers::info',
  REGISTER_WORKER: 'engine::workers::register',
} as const

/** Engine trigger types */
export const EngineTriggers = {
  FUNCTIONS_AVAILABLE: 'engine::functions-available',
} as const

/** Connection state for the III WebSocket */
export type IIIConnectionState = 'disconnected' | 'connecting' | 'connected' | 'reconnecting' | 'failed'

/** Configuration for WebSocket reconnection behavior */
export interface IIIReconnectionConfig {
  /** Starting delay in milliseconds (default: 1000ms) */
  initialDelayMs: number
  /** Maximum delay cap in milliseconds (default: 30000ms) */
  maxDelayMs: number
  /** Exponential backoff multiplier (default: 2) */
  backoffMultiplier: number
  /** Random jitter factor 0-1 (default: 0.3) */
  jitterFactor: number
  /** Maximum retry attempts, -1 for infinite (default: -1) */
  maxRetries: number
}

/** Default reconnection configuration */
export const DEFAULT_BRIDGE_RECONNECTION_CONFIG: IIIReconnectionConfig = {
  initialDelayMs: 1000,
  maxDelayMs: 30000,
  backoffMultiplier: 2,
  jitterFactor: 0.3,
  maxRetries: -1,
}

/** Default invocation timeout in milliseconds */
export const DEFAULT_INVOCATION_TIMEOUT_MS = 30000
